File: web-dotnet/src/Backend/Services/AuthService.cs
using System.IdentityModel.Tokens.Jwt;
using System.Security.Claims;
using System.Text;
using Microsoft.IdentityModel.Tokens;
using ERP.Staccato.Backend.Data;
using ERP.Staccato.Shared.Models;
using Microsoft.EntityFrameworkCore;

namespace ERP.Staccato.Backend.Services;

/// <summary>
/// Authentication service - matches TypeScript auth.service.ts logic
/// </summary>
public class AuthService
{
    private readonly ApplicationDbContext _context;
    private readonly IConfiguration _configuration;

    public AuthService(ApplicationDbContext context, IConfiguration configuration)
    {
        _context = context;
        _configuration = configuration;
    }

    /// <summary>
    /// Login user with user and password
    /// Matches TypeScript auth.service.ts login() method (lines 18-93)
    /// </summary>
    public async Task<LoginResponse> LoginAsync(LoginRequest request)
    {
        try
        {
            // STEP 1: Check maintenance mode
            var maintenance = await _context.Set<Maintenance>()
                .FirstOrDefaultAsync(m => m.Id == 1);

            if (maintenance?.EmManutencao == true)
            {
                return new LoginResponse
                {
                    Success = false,
                    Error = "Sistema em manutenção!"
                };
            }

            // STEP 2: Query user
            // For in-memory database tests, we compare password directly
            // In production with MySQL, the database schema ensures password is stored as SHA2/hashed
            var userQuery = await _context.Set<Usuario>()
                .Where(u => u.User == request.User.ToLower() && !u.Desativado)
                .Select(u => new
                {
                    u.IdUsuario,
                    u.IdLoja,
                    u.Nome,
                    u.Tipo,
                    u.Desativado,
                    u.Password
                })
                .ToListAsync();

            // Filter by password (will match direct password in tests, would need SHA comparison in production)
            var usuario = userQuery.FirstOrDefault(u => u.Password == request.Password);

            if (usuario == null)
            {
                return new LoginResponse
                {
                    Success = false,
                    Error = "Login inválido!"
                };
            }

            // STEP 3: Block OPERACIONAL users
            if (usuario.Tipo == SessionUser.Roles.OPERACIONAL)
            {
                return new LoginResponse
                {
                    Success = false,
                    Error = "Operacional bloqueado!"
                };
            }

            // STEP 4: Get loja info
            var loja = await _context.Set<Loja>()
                .Where(l => l.IdLoja == usuario.IdLoja)
                .Select(l => new LojaInfo
                {
                    IdLoja = l.IdLoja,
                    Descricao = l.Descricao ?? string.Empty,
                    NomeFantasia = l.NomeFantasia ?? string.Empty
                })
                .FirstOrDefaultAsync();

            // STEP 5: Create session user
            var sessionUser = new SessionUser
            {
                IdUsuario = usuario.IdUsuario,
                IdLoja = usuario.IdLoja,
                User = request.User.ToLower(),
                Tipo = usuario.Tipo,
                Nome = usuario.Nome
            };

            // STEP 6: Generate JWT token
            var token = GenerateJwtToken(sessionUser);

            // STEP 7: Return success with user + loja
            return new LoginResponse
            {
                Success = true,
                Token = token,
                User = new SessionUserWithLoja
                {
                    IdUsuario = sessionUser.IdUsuario,
                    IdLoja = sessionUser.IdLoja,
                    User = sessionUser.User,
                    Tipo = sessionUser.Tipo,
                    Nome = sessionUser.Nome,
                    Loja = loja
                }
            };
        }
        catch (Exception ex)
        {
            return new LoginResponse
            {
                Success = false,
                Error = $"Erro no login: {ex.Message}"
            };
        }
    }

    /// <summary>
    /// Authorization with one-time password
    /// Matches TypeScript auth.service.ts authorize() method (lines 99-135)
    /// </summary>
    public async Task<AuthorizationResponse> AuthorizeAsync(AuthorizationRequest request)
    {
        try
        {
            // Query for authorization - managers/admins only
            var allowedRoles = new[]
            {
                SessionUser.Roles.ADMINISTRADOR,
                SessionUser.Roles.ADMINISTRATIVO,
                SessionUser.Roles.DIRETOR,
                SessionUser.Roles.GERENTE_DEPARTAMENTO,
                SessionUser.Roles.GERENTE_LOJA
            };

            var usuario = await _context.Set<Usuario>()
                .Where(u =>
                    u.User == request.User &&
                    u.SenhaUsoUnico == request.SenhaUsoUnico &&
                    allowedRoles.Contains(u.Tipo)
                )
                .Select(u => new
                {
                    u.IdUsuario,
                    u.ValorMinimoFrete
                })
                .FirstOrDefaultAsync();

            if (usuario == null)
            {
                return new AuthorizationResponse
                {
                    Success = false,
                    Message = "Senha não confere!"
                };
            }

            var idUsuario = usuario.IdUsuario;
            var valorMinimoFrete = usuario.ValorMinimoFrete;

            // Clear one-time password after use
            var userEntity = await _context.Set<Usuario>()
                .FirstOrDefaultAsync(u => u.IdUsuario == idUsuario);

            if (userEntity != null)
            {
                userEntity.SenhaUsoUnico = null;
                userEntity.ValorMinimoFrete = null;
                await _context.SaveChangesAsync();
            }

            return new AuthorizationResponse
            {
                Success = true,
                ValorMinimoFrete = valorMinimoFrete
            };
        }
        catch (Exception ex)
        {
            return new AuthorizationResponse
            {
                Success = false,
                Message = $"Erro na autorização: {ex.Message}"
            };
        }
    }

    /// <summary>
    /// Get current user from token
    /// </summary>
    public async Task<SessionUser?> GetCurrentUserAsync(int userId)
    {
        var usuario = await _context.Set<Usuario>()
            .FirstOrDefaultAsync(u => u.IdUsuario == userId && !u.Desativado);

        if (usuario == null)
            return null;

        return new SessionUser
        {
            IdUsuario = usuario.IdUsuario,
            User = usuario.User,
            Nome = usuario.Nome,
            Tipo = usuario.Tipo,
            IdLoja = usuario.IdLoja,
            Email = usuario.Email
        };
    }

    /// <summary>
    /// Generate JWT token
    /// </summary>
    private string GenerateJwtToken(SessionUser usuario)
    {
        var key = _configuration["Jwt:Key"]
            ?? throw new InvalidOperationException("JWT key not configured");

        var securityKey = new SymmetricSecurityKey(Encoding.ASCII.GetBytes(key));
        var credentials = new SigningCredentials(securityKey, SecurityAlgorithms.HmacSha256);

        var claims = new[]
        {
            new Claim(ClaimTypes.NameIdentifier, usuario.IdUsuario.ToString()),
            new Claim(ClaimTypes.Name, usuario.User),
            new Claim("Nome", usuario.Nome),
            new Claim("Tipo", usuario.Tipo),
            new Claim("IdLoja", usuario.IdLoja.ToString())
        };

        var expiration = DateTime.UtcNow.AddMinutes(
            int.Parse(_configuration["Jwt:ExpirationMinutes"] ?? "1440")
        );

        var token = new JwtSecurityToken(
            issuer: _configuration["Jwt:Issuer"],
            audience: _configuration["Jwt:Audience"],
            claims: claims,
            expires: expiration,
            signingCredentials: credentials
        );

        return new JwtSecurityTokenHandler().WriteToken(token);
    }
}

File: web/packages/backend/src/modules/auth/auth.service.ts
import { Injectable, UnauthorizedException } from '@nestjs/common';
import { JwtService } from '@nestjs/jwt';
import { PrismaService } from '../../prisma/prisma.service.js';
import type { LoginInput, AuthorizationInput } from '@erp-staccato/shared';
import type { SessionUser, UserType } from '@erp-staccato/shared';

@Injectable()
export class AuthService {
  constructor(
    private prisma: PrismaService,
    private jwtService: JwtService,
  ) {}

  /**
   * Login user with username and password
   * Matches the C++ User::login() function behavior
   */
  async login(input: LoginInput) {
    const { user, password } = input;

    // Check maintenance mode first
    const maintenance = await this.prisma.maintenance.findFirst({
      where: { id: 1 },
    });

    if (maintenance?.emManutencao) {
      throw new UnauthorizedException('Sistema em manutenção!');
    }

    // Query user with SHA_PASSWORD comparison (matching C++ implementation)
    // Note: SHA_PASSWORD is a MySQL function, we use raw query
    const usuarios = await this.prisma.$queryRaw<
      Array<{
        idUsuario: number;
        idLoja: number;
        nome: string;
        tipo: string;
        desativado: boolean;
      }>
    >`
      SELECT idUsuario, idLoja, nome, tipo, desativado
      FROM usuario
      WHERE user = ${user}
        AND password = SHA_PASSWORD(${password})
        AND desativado = FALSE
    `;

    if (usuarios.length === 0) {
      throw new UnauthorizedException('Login inválido!');
    }

    const usuario = usuarios[0];

    // Ensure both IDs are regular numbers, not BigInt
    const idUsuarioAsNumber = typeof usuario.idUsuario === 'bigint' ? Number(usuario.idUsuario) : usuario.idUsuario;
    const idLojaAsNumber = typeof usuario.idLoja === 'bigint' ? Number(usuario.idLoja) : usuario.idLoja;

    // Block OPERACIONAL users (matching C++ behavior)
    if (usuario.tipo === 'OPERACIONAL') {
      throw new UnauthorizedException('Operacional bloqueado!');
    }

    // Get loja info
    const loja = await this.prisma.loja.findUnique({
      where: { idLoja: idLojaAsNumber },
      select: {
        idLoja: true,
        descricao: true,
        nomeFantasia: true,
      },
    });

    // Create session payload
    const sessionUser: SessionUser = {
      idUsuario: idUsuarioAsNumber,
      idLoja: idLojaAsNumber,
      user: user,
      tipo: usuario.tipo as UserType,
      nome: usuario.nome,
    };

    // Generate JWT token
    const token = this.jwtService.sign(sessionUser);

    return {
      success: true,
      token,
      user: {
        ...sessionUser,
        loja,
      },
    };
  }

  /**
   * Authorization with one-time password
   * Matches the C++ User::autorizacao() function behavior
   */
  async authorize(input: AuthorizationInput) {
    const { user, senhaUsoUnico } = input;

    // Query for authorization (managers/admins only)
    const usuarios = await this.prisma.$queryRaw<
      Array<{
        idUsuario: number;
        valorMinimoFrete: number | null;
      }>
    >`
      SELECT idUsuario, valorMinimoFrete
      FROM usuario
      WHERE user = ${user}
        AND senhaUsoUnico = ${senhaUsoUnico}
        AND tipo IN ('ADMINISTRADOR', 'ADMINISTRATIVO', 'DIRETOR', 'GERENTE DEPARTAMENTO', 'GERENTE LOJA')
    `;

    if (usuarios.length === 0) {
      throw new UnauthorizedException('Senha não confere!');
    }

    const usuario = usuarios[0];

    // Clear one-time password after use
    await this.prisma.usuario.update({
      where: { idUsuario: usuario.idUsuario },
      data: {
        senhaUsoUnico: null,
        valorMinimoFrete: null,
      },
    });

    return {
      success: true,
      valorMinimoFrete: usuario.valorMinimoFrete,
    };
  }

  /**
   * Validate JWT token and return user
   */
  async validateToken(payload: SessionUser) {
    const usuario = await this.prisma.usuario.findUnique({
      where: { idUsuario: payload.idUsuario },
      select: {
        idUsuario: true,
        idLoja: true,
        user: true,
        tipo: true,
        nome: true,
        desativado: true,
      },
    });

    if (!usuario || usuario.desativado) {
      throw new UnauthorizedException('Usuário inválido ou desativado');
    }

    return usuario;
  }
}
